File: packages/python/emgran_a2a/connection.py
"""
A2A Connection - Manages connections between agents
"""

from typing import Dict, Any
from datetime import datetime


class A2AConnection:
    """
    A2A Connection
    
    Manages a connection between two agents
    """
    
    def __init__(self, local_agent: Dict[str, str], remote_agent: Dict[str, str]):
        """
        Initialize A2A Connection
        
        Args:
            local_agent: Local agent information
            remote_agent: Remote agent information
        """
        self.id = f'conn_{int(datetime.now().timestamp())}'
        self.local_agent = local_agent
        self.remote_agent = remote_agent
        self.status = 'active'
        self.created_at = datetime.now().isoformat()
        self.last_activity = datetime.now().isoformat()
    
    async def send(self, data: Any) -> Dict[str, Any]:
        """
        Send data to remote agent
        
        Args:
            data: Data to send
            
        Returns:
            Send result
        """
        # TODO: Implement actual data sending logic
        self._update_activity()
        return {
            'success': True,
            'operation': 'send',
            'connection_id': self.id,
            'remote_agent': self.remote_agent['agent_id'],
            'data_sent': data,
            'timestamp': datetime.now().isoformat(),
        }
    
    async def receive(self) -> Dict[str, Any]:
        """
        Receive data from remote agent
        
        Returns:
            Receive result with data
        """
        # TODO: Implement actual data receiving logic
        self._update_activity()
        return {
            'success': True,
            'operation': 'receive',
            'connection_id': self.id,
            'remote_agent': self.remote_agent['agent_id'],
            'data_received': {},  # Mock data
            'timestamp': datetime.now().isoformat(),
        }
    
    async def close(self) -> None:
        """
        Close the connection
        """
        self.status = 'closed'
        self._update_activity()
    
    def _update_activity(self) -> None:
        """Update last activity timestamp"""
        self.last_activity = datetime.now().isoformat()
    
    def to_dict(self) -> Dict[str, Any]:
        """
        Convert connection to dictionary
        
        Returns:
            Connection information as dict
        """
        return {
            'id': self.id,
            'local_agent': self.local_agent,
            'remote_agent': self.remote_agent,
            'status': self.status,
            'created_at': self.created_at,
            'last_activity': self.last_activity,
        }

File: packages/typescript/src/services/connection.ts
import { A2AParticipant, A2AConnection } from '../types';

/**
 * A2A Connection Service
 * 
 * Manages connections between agents
 */
export class A2AConnectionService {
  private localAgent: A2AParticipant;
  private connections: Map<string, A2AConnection>;

  constructor(localAgent: A2AParticipant) {
    this.localAgent = localAgent;
    this.connections = new Map();
  }

  /**
   * Connect to another agent
   */
  async connect(local: A2AParticipant, remote: A2ADiscoveryResult): Promise<A2AConnection> {
    const connection: A2AConnection = {
      id: `conn_${Date.now()}`,
      local_agent: local,
      remote_agent: {
        did: remote.did,
        agent_id: remote.agent_id,
        endpoint: remote.endpoint,
      },
      status: 'active',
      created_at: new Date().toISOString(),
      last_activity: new Date().toISOString(),
    };

    this.connections.set(connection.id, connection);
    return connection;
  }

  /**
   * Close a connection
   */
  async close(connectionId: string): Promise<void> {
    const connection = this.connections.get(connectionId);
    if (connection) {
      connection.status = 'closed';
      this.connections.set(connectionId, connection);
    }
  }

  /**
   * Get all active connections
   */
  getConnections(): A2AConnection[] {
    return Array.from(this.connections.values());
  }
}
